rodbus/src/tcp/tls/server.rs

```diff
@@ -0,0 +1,325 @@
+                            .map_err(|err| err.to_string())
+                    })
+                } else {
+                    return Err(TlsError::InvalidPeerCertificate(io::Error::new(
+                        ErrorKind::InvalidData,
+                        "no peer certificate",
+                    )));
+                }
+            }
+        };
+
+        Ok(TlsServerConfig { config_builder })
+    }
+
+    fn build(&self, role_container: RoleContainer) -> Result<Arc<rustls::ServerConfig>, String> {
+        Ok(Arc::new((self.config_builder)(role_container)?))
+    }
+
+    pub(crate) async fn handle_connection(
+        &mut self,
+        socket: TcpStream,
+        level: PhysDecodeLevel,
+        auth_handler: Arc<dyn AuthorizationHandler>,
+    ) -> Result<(PhysLayer, SessionAuthentication), String> {
+        let role_container = Arc::new(Mutex::new(None));
+        let tls_config = self.build(role_container.clone())?;
+
+        let connector = tokio_rustls::TlsAcceptor::from(tls_config);
+        match connector.accept(socket).await {
+            Err(err) => Err(format!("failed to establish TLS session: {}", err)),
+            Ok(stream) => {
+                let layer = PhysLayer::new_tls(tokio_rustls::TlsStream::from(stream), level);
+                let role = role_container
+                    .lock()
+                    .unwrap()
+                    .clone()
+                    .ok_or_else(|| "client did not present Modbus role".to_string())?;
+
+                Ok((
+                    layer,
+                    SessionAuthentication::Authenticated(auth_handler, role),
+                ))
+            }
+        }
+    }
+}
+
+struct CaChainClientCertVerifier {
+    inner: Arc<dyn rustls::server::ClientCertVerifier>,
+    role_container: RoleContainer,
+}
+
+impl CaChainClientCertVerifier {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(
+        roots: rustls::RootCertStore,
+        role_container: RoleContainer,
+    ) -> Arc<dyn rustls::server::ClientCertVerifier> {
+        let inner = AllowAnyAuthenticatedClient::new(roots);
+        Arc::new(CaChainClientCertVerifier {
+            inner,
+            role_container,
+        })
+    }
+}
+
+impl rustls::server::ClientCertVerifier for CaChainClientCertVerifier {
+    fn offer_client_auth(&self) -> bool {
+        // Client must authenticate itself, so we better offer the authentication!
+        true
+    }
+
+    fn client_auth_mandatory(&self) -> Option<bool> {
+        // Client must authenticate itself
+        Some(true)
+    }
+
+    fn client_auth_root_subjects(&self) -> Option<rustls::DistinguishedNames> {
+        self.inner.client_auth_root_subjects()
+    }
+
+    fn verify_client_cert(
+        &self,
+        end_entity: &rustls::Certificate,
+        intermediates: &[rustls::Certificate],
+        now: std::time::SystemTime,
+    ) -> Result<rustls::server::ClientCertVerified, rustls::Error> {
+        self.inner
```

Comment:
It is supported for the client. It doesn't make much sense on the server side, since we don't validate the client name at all (DNP3 is different though). The relevant code is here:
https://github.com/stepfunc/rodbus/blob/1c6e9f1c8f087353553362c18a5ce8149850f7ae/rodbus/src/tcp/tls/mod.rs#L91-L163